From 65cdf080450e23ce6fd0e463d47f5d0c306b88e1 Mon Sep 17 00:00:00 2001
From: Shunqing Chen <csq@rock-chips.com>
Date: Sun, 19 Sep 2021 17:34:51 +0800
Subject: [PATCH 103/232] misc: rk628-for-all: add irq support for hdmirx

plugin-det-gpios: hdmirx plugin detect
hpd-output-inverted: hpd level inverted

Signed-off-by: Shunqing Chen <csq@rock-chips.com>
Change-Id: I3e6dceb3e7b13ce0f51b24e60e71abe9804818f3
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/misc/rk628/rk628.c        | 76 ++++++++++++++++++++++++++--
 drivers/misc/rk628/rk628.h        |  3 ++
 drivers/misc/rk628/rk628_hdmirx.c | 84 +++++++++++++++++++++++++++----
 drivers/misc/rk628/rk628_hdmirx.h | 14 ++++++
 4 files changed, 164 insertions(+), 13 deletions(-)
 mode change 100755 => 100644 drivers/misc/rk628/rk628.c

diff --git a/drivers/misc/rk628/rk628.c b/drivers/misc/rk628/rk628.c
old mode 100755
new mode 100644
index bd812ddf20f465..d95f133ec199c6
--- a/drivers/misc/rk628/rk628.c
+++ b/drivers/misc/rk628/rk628.c
@@ -8,6 +8,7 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/interrupt.h>
+#include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio/consumer.h>
 #include <linux/regmap.h>
@@ -340,11 +341,31 @@ static void rk628_display_work(struct work_struct *work)
 	}
 
 #ifdef RK628_HDMI_IN
-	queue_delayed_work(rk628->monitor_wq, &rk628->delay_work,
+	if (!rk628->plugin_det_gpio)
+		queue_delayed_work(rk628->monitor_wq, &rk628->delay_work,
 				   msecs_to_jiffies(2000));
+	else
+		rk628_hdmirx_enable_interrupts(rk628, true);
 #endif
 }
 
+#ifdef RK628_HDMI_IN
+static irqreturn_t rk628_hdmirx_plugin_irq(int irq, void *dev_id)
+{
+	struct rk628 *rk628 = dev_id;
+
+	/* control hpd after 50ms */
+	schedule_delayed_work(&rk628->delay_work, HZ / 20);
+
+	rk628_hdmirx_enable_interrupts(rk628, false);
+	/* clear interrupts */
+	rk628_i2c_write(rk628, HDMI_RX_MD_ICLR, 0xffffffff);
+	rk628_i2c_write(rk628, HDMI_RX_PDEC_ICLR, 0xffffffff);
+	rk628_i2c_write(rk628, GRF_INTR0_CLR_EN, 0x01000100);
+	return IRQ_HANDLED;
+}
+#endif
+
 static int
 rk628_i2c_probe(struct i2c_client *client, const struct i2c_device_id *id)
 {
@@ -352,6 +373,10 @@ rk628_i2c_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	struct device_node *backlight;
 	struct rk628 *rk628;
 	int i, ret;
+#ifdef RK628_HDMI_IN
+	int err;
+	unsigned long irq_flags;
+#endif
 
 	dev_info(dev, "RK628 misc driver version: %s\n", DRIVER_VERSION);
 
@@ -362,6 +387,7 @@ rk628_i2c_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	rk628->dev = dev;
 	rk628->client = client;
 	i2c_set_clientdata(client, rk628);
+	rk628->hdmirx_irq = client->irq;
 
 	rk628->soc_24M = devm_clk_get(dev, "soc_24M");
 	if (rk628->soc_24M == ERR_PTR(-ENOENT))
@@ -390,6 +416,14 @@ rk628_i2c_probe(struct i2c_client *client, const struct i2c_device_id *id)
 		return ret;
 	}
 
+	rk628->plugin_det_gpio = devm_gpiod_get_optional(dev, "plugin-det",
+						    GPIOD_IN);
+	if (IS_ERR(rk628->plugin_det_gpio)) {
+		dev_err(rk628->dev, "failed to get hdmirx det gpio\n");
+		ret = PTR_ERR(rk628->plugin_det_gpio);
+		return ret;
+	}
+
 	rk628->panel = devm_kzalloc(dev, sizeof(struct panel_simple), GFP_KERNEL);
 	if (!rk628->panel)
 		return -ENOMEM;
@@ -454,10 +488,44 @@ rk628_i2c_probe(struct i2c_client *client, const struct i2c_device_id *id)
 #ifdef RK628_CSI_OUT
 	rk628_csi_init(rk628);
 #endif
-	rk628_display_enable(rk628);
 
-	queue_delayed_work(rk628->monitor_wq, &rk628->delay_work,
-			   msecs_to_jiffies(3000));
+#ifdef RK628_HDMI_IN
+	if (rk628->plugin_det_gpio) {
+		rk628->plugin_irq = gpiod_to_irq(rk628->plugin_det_gpio);
+		if (rk628->plugin_irq < 0) {
+			dev_err(rk628->dev, "failed to get plugin det irq\n");
+			err = rk628->plugin_irq;
+			return err;
+		}
+
+		err = devm_request_threaded_irq(dev, rk628->plugin_irq, NULL,
+				rk628_hdmirx_plugin_irq, IRQF_TRIGGER_FALLING |
+				IRQF_TRIGGER_RISING | IRQF_ONESHOT, "rk628_hdmirx", rk628);
+		if (err) {
+			dev_err(rk628->dev, "failed to register plugin det irq (%d)\n", err);
+			return err;
+		}
+
+		if (rk628->hdmirx_irq) {
+			irq_flags = irqd_get_trigger_type(irq_get_irq_data(rk628->hdmirx_irq));
+			dev_dbg(rk628->dev, "cfg hdmirx irq, flags: %lu!\n", irq_flags);
+			err = devm_request_threaded_irq(dev, rk628->hdmirx_irq, NULL,
+					rk628_hdmirx_plugin_irq, irq_flags |
+					IRQF_ONESHOT, "rk628", rk628);
+			if (err) {
+				dev_err(rk628->dev, "request rk628 irq failed! err:%d\n",
+						err);
+				return err;
+			}
+			/* hdmirx int en */
+			rk628_i2c_write(rk628, GRF_INTR0_EN, 0x01000100);
+		}
+	} else {
+		queue_delayed_work(rk628->monitor_wq, &rk628->delay_work,
+				    msecs_to_jiffies(3000));
+	}
+#endif
+	rk628_display_enable(rk628);
 
 	return 0;
 }
diff --git a/drivers/misc/rk628/rk628.h b/drivers/misc/rk628/rk628.h
index e85dae0242e61d..7a75978ab6809d 100644
--- a/drivers/misc/rk628/rk628.h
+++ b/drivers/misc/rk628/rk628.h
@@ -259,6 +259,9 @@ struct rk628 {
 	struct regmap *regmap[RK628_DEV_MAX];
 	struct gpio_desc *reset_gpio;
 	struct gpio_desc *enable_gpio;
+	struct gpio_desc *plugin_det_gpio;
+	int plugin_irq;
+	int hdmirx_irq;
 	struct clk *soc_24M;
 	struct workqueue_struct *monitor_wq;
 	struct delayed_work delay_work;
diff --git a/drivers/misc/rk628/rk628_hdmirx.c b/drivers/misc/rk628/rk628_hdmirx.c
index 29b7752894cc62..5719970b55771e 100644
--- a/drivers/misc/rk628/rk628_hdmirx.c
+++ b/drivers/misc/rk628/rk628_hdmirx.c
@@ -3,6 +3,9 @@
  *
  * Copyright (c) 2020 Fuzhou Rockchip Electronics Co., Ltd\
  */
+#include <linux/gpio/consumer.h>
+#include <linux/of_device.h>
+
 #include "rk628.h"
 #include "rk628_combrxphy.h"
 #include "rk628_config.h"
@@ -71,6 +74,7 @@ struct rk628_hdmirx {
 	u32 input_format;
 	u32 fs_audio;
 	bool audio_present;
+	bool hpd_output_inverted;
 };
 
 const static struct rk628_hdmi_mode supported_modes[] = {
@@ -128,22 +132,51 @@ static void rk628_hdmirx_video_unmute(struct rk628 *rk628, u8 unmute)
 	rk628_i2c_update_bits(rk628, HDMI_RX_DMI_DISABLE_IF, VID_ENABLE_MASK, VID_ENABLE(unmute));
 }
 
+static void rk628_hdmirx_hpd_ctrl(struct rk628 *rk628, bool en)
+{
+	u8 en_level, set_level;
+	struct rk628_hdmirx *hdmirx = rk628->hdmirx;
+
+	dev_dbg(rk628->dev, "%s: %sable, hpd invert:%d\n", __func__,
+			en ? "en" : "dis", hdmirx->hpd_output_inverted);
+	en_level = hdmirx->hpd_output_inverted ? 0 : 1;
+	set_level = en ? en_level : !en_level;
+	rk628_i2c_update_bits(rk628, HDMI_RX_HDMI_SETUP_CTRL,
+			HOT_PLUG_DETECT_MASK, HOT_PLUG_DETECT(set_level));
+}
+
 static void rk628_hdmirx_disable_edid(struct rk628 *rk628)
 {
-	rk628_i2c_update_bits(rk628, HDMI_RX_HDMI_SETUP_CTRL, HOT_PLUG_DETECT_MASK,
-	     HOT_PLUG_DETECT(0));
+	rk628_hdmirx_hpd_ctrl(rk628, false);
 	rk628_hdmirx_video_unmute(rk628, 0);
 }
 
 static void rk628_hdmirx_enable_edid(struct rk628 *rk628)
 {
-	rk628_i2c_update_bits(rk628, HDMI_RX_HDMI_SETUP_CTRL, HOT_PLUG_DETECT_MASK,
-	     HOT_PLUG_DETECT(1));
+	rk628_hdmirx_hpd_ctrl(rk628, true);
 }
 
-static int tx_5v_power_present(void)
+static int tx_5v_power_present(struct rk628 *rk628)
 {
-	return 1;
+	bool ret;
+	int val, i, cnt;
+
+	/* Direct Mode */
+	if (!rk628->plugin_det_gpio)
+		return 1;
+
+	cnt = 0;
+	for (i = 0; i < 5; i++) {
+		val = gpiod_get_value(rk628->plugin_det_gpio);
+		if (val > 0)
+			cnt++;
+		usleep_range(500, 600);
+	}
+
+	ret = (cnt >= 3) ? 1 : 0;
+	dev_dbg(rk628->dev, "%s: %d\n", __func__, ret);
+
+	return ret;
 }
 
 static int rk628_hdmirx_set_edid(struct rk628 *rk628)
@@ -506,6 +539,9 @@ static bool rk628_hdmirx_status_change(struct rk628 *rk628)
 	hact = val & 0xffff;
 	rk628_i2c_read(rk628, HDMI_RX_MD_VAL, &val);
 	vact = val & 0xffff;
+	if (!rk628->plugin_det_gpio && !hact && !vact)
+		return true;
+
 	if (hact != hdmirx->mode.hdisplay || vact != hdmirx->mode.vdisplay) {
 		printk("new: hdisplay=%d, vdisplay=%d\n", hact, vact);
 		printk("old: hdisplay=%d, vdisplay=%d\n", hdmirx->mode.hdisplay, hdmirx->mode.vdisplay);
@@ -522,12 +558,16 @@ static bool rk628_hdmirx_status_change(struct rk628 *rk628)
 static int rk628_hdmirx_init(struct rk628 *rk628)
 {
 	struct rk628_hdmirx *hdmirx;
+	struct device *dev = rk628->dev;
 
 	hdmirx = devm_kzalloc(rk628->dev, sizeof(*hdmirx), GFP_KERNEL);
 	if (!hdmirx)
 		return -ENOMEM;
 	rk628->hdmirx = hdmirx;
 
+	hdmirx->hpd_output_inverted = of_property_read_bool(dev->of_node,
+		"hpd-output-inverted");
+
 	/* HDMIRX IOMUX */
 	rk628_i2c_write(rk628, GRF_GPIO1AB_SEL_CON, HIWORD_UPDATE(0x7, 10, 8));
 	//i2s pinctrl
@@ -545,6 +585,32 @@ static int rk628_hdmirx_init(struct rk628 *rk628)
 	return 0;
 }
 
+void rk628_hdmirx_enable_interrupts(struct rk628 *rk628, bool en)
+{
+	u32 pdec_ien, md_ien;
+	u32 pdec_mask = 0, md_mask = 0;
+
+	pdec_mask |= AVI_RCV_ENSET;
+	md_mask = VACT_LIN_ENSET | HACT_PIX_ENSET | HS_CLK_ENSET |
+		  DE_ACTIVITY_ENSET | VS_ACT_ENSET | HS_ACT_ENSET;
+	dev_dbg(rk628->dev, "%s: %sable\n", __func__, en ? "en" : "dis");
+	/* clr irq */
+	rk628_i2c_write(rk628, HDMI_RX_MD_ICLR, md_mask);
+	rk628_i2c_write(rk628, HDMI_RX_PDEC_ICLR, pdec_mask);
+	if (en) {
+		rk628_i2c_write(rk628, HDMI_RX_MD_IEN_SET, md_mask);
+		rk628_i2c_write(rk628, HDMI_RX_PDEC_IEN_SET, pdec_mask);
+	} else {
+		rk628_i2c_write(rk628, HDMI_RX_MD_IEN_CLR, md_mask);
+		rk628_i2c_write(rk628, HDMI_RX_PDEC_IEN_CLR, pdec_mask);
+		rk628_i2c_write(rk628, HDMI_RX_AUD_FIFO_IEN_CLR, 0x1f);
+	}
+	usleep_range(5000, 5000);
+	rk628_i2c_read(rk628, HDMI_RX_MD_IEN, &md_ien);
+	rk628_i2c_read(rk628, HDMI_RX_PDEC_IEN, &pdec_ien);
+	dev_dbg(rk628->dev, "%s MD_IEN:%#x, PDEC_IEN:%#x\n", __func__, md_ien, pdec_ien);
+}
+
 int rk628_hdmirx_enable(struct rk628 *rk628)
 {
 	int ret;
@@ -557,7 +623,7 @@ int rk628_hdmirx_enable(struct rk628 *rk628)
 	}
 
 	hdmirx = rk628->hdmirx;
-	if (tx_5v_power_present()) {
+	if (tx_5v_power_present(rk628)) {
 		hdmirx->plugin = true;
 		rk628_hdmirx_enable_edid(rk628);
 		rk628_hdmirx_ctrl_enable(rk628);
@@ -593,7 +659,7 @@ void rk628_hdmirx_disable(struct rk628 *rk628)
 	}
 
 	hdmirx = rk628->hdmirx;
-	if (!tx_5v_power_present()) {
+	if (!tx_5v_power_present(rk628)) {
 		hdmirx->plugin = false;
 		rk628_hdmirx_disable_edid(rk628);
 		rk628_i2c_update_bits(rk628, GRF_SYSTEM_CON0, SW_I2S_DATA_OEN_MASK, SW_I2S_DATA_OEN(1));
@@ -609,7 +675,7 @@ int rk628_hdmirx_detect(struct rk628 *rk628)
 	if (!hdmirx)
 		return HDMIRX_PLUGOUT;
 
-	if (tx_5v_power_present()) {
+	if (tx_5v_power_present(rk628)) {
 		ret |= HDMIRX_PLUGIN;
 		if (!hdmirx->plugin)
 			ret |= HDMIRX_CHANGED;
diff --git a/drivers/misc/rk628/rk628_hdmirx.h b/drivers/misc/rk628/rk628_hdmirx.h
index 29ef2bcf14f800..d83025a782c59f 100644
--- a/drivers/misc/rk628/rk628_hdmirx.h
+++ b/drivers/misc/rk628/rk628_hdmirx.h
@@ -461,6 +461,13 @@
 #define HDMI_RX_HDMI2_ISET			HDMIRX_REG(0x0f74)
 #define HDMI_RX_PDEC_IEN_CLR			HDMIRX_REG(0x0f78)
 #define HDMI_RX_PDEC_IEN_SET			HDMIRX_REG(0x0f7c)
+#define ACR_N_CHG_IEN				BIT(23)
+#define ACR_CTS_CHG_IEN				BIT(22)
+#define GCP_AV_MUTE_CHG_ENSET			BIT(21)
+#define AIF_RCV_ENSET				BIT(19)
+#define AVI_RCV_ENSET				BIT(18)
+#define GCP_RCV_ENSET				BIT(16)
+#define AMP_RCV_ENSET				BIT(14)
 #define HDMI_RX_PDEC_ISTS			HDMIRX_REG(0x0f80)
 #define AVI_RCV_ISTS				BIT(18)
 #define HDMI_RX_PDEC_IEN			HDMIRX_REG(0x0f84)
@@ -481,6 +488,12 @@
 #define HDMI_RX_AUD_FIFO_ISET			HDMIRX_REG(0x0fbc)
 #define HDMI_RX_MD_IEN_CLR			HDMIRX_REG(0x0fc0)
 #define HDMI_RX_MD_IEN_SET			HDMIRX_REG(0x0fc4)
+#define VACT_LIN_ENSET				BIT(9)
+#define HACT_PIX_ENSET				BIT(6)
+#define HS_CLK_ENSET				BIT(5)
+#define DE_ACTIVITY_ENSET			BIT(2)
+#define VS_ACT_ENSET				BIT(1)
+#define HS_ACT_ENSET				BIT(0)
 #define HDMI_RX_MD_ISTS				HDMIRX_REG(0x0fc8)
 #define HDMI_RX_MD_IEN				HDMIRX_REG(0x0fcc)
 #define HDMI_RX_MD_ICLR				HDMIRX_REG(0x0fd0)
@@ -611,6 +624,7 @@
 #define HDMIRX_CHANGED				BIT(2)
 #define HDMIRX_NOSIGNAL				BIT(3)
 
+void rk628_hdmirx_enable_interrupts(struct rk628 *rk628, bool en);
 int rk628_hdmirx_enable(struct rk628 *rk628);
 void rk628_hdmirx_disable(struct rk628 *rk628);
 int rk628_hdmirx_detect(struct rk628 *rk628);
-- 
2.17.1

